mrmustard/lab_dev/sampler.py

```diff
@@ -0,0 +1,192 @@
+            meas_op = self.povms[self.meas_outcomes.index(unique_sample)].on([initial_mode])
+            reduced_state = (state >> meas_op.dual).normalize()
+            samples = self.sample(reduced_state, counts)
+            for sample in samples:
+                ret.append(np.append([unique_sample], sample))
+        return np.array(ret)
+
+    def sample_prob_dist(
+        self, state: State, n_samples: int = 1000, seed: int | None = None
+    ) -> np.ndarray:
+        r"""
+        Samples a state by computing the probability distribution.
+
+        Args:
+            state: The state to sample.
+            n_samples: The number of samples to generate.
+            seed: An optional seed for random sampling.
+        """
+        rng = np.random.default_rng(seed) if seed else settings.rng
+        return rng.choice(
+            a=list(product(self.meas_outcomes, repeat=len(state.modes))),
+            p=self.probabilities(state),
+            size=n_samples,
+        )
+
+    def _validate_probs(self, probs: Sequence[float], dx: float, atol: float) -> Sequence[float]:
+        r"""
+        Validates that the given probability distribution sums to `1.0` within some
+        tolerance and returns a renormalized probability distribution to account for
+        small numerical errors.
+
+        Args:
+            probs: The probability distribution to validate.
+            dx: The uniform differential for the probability distribution.
+            atol: The absolute tolerance to validate with.
+        """
+        atol = atol or settings.ATOL
+        probs_dx = probs * dx
+        prob_sum = sum(probs_dx)
+        if not math.allclose(prob_sum, 1, atol):
+            raise ValueError(f"Probabilities sum to {prob_sum} and not 1.0.")
+        return math.real(probs_dx / prob_sum)
+
+
+class PNRSampler(Sampler):
+    r"""
+    A sampler for photon-number resolving (PNR) detectors.
+
+    Args:
+        cutoff: The photon number cutoff.
+    """
+
+    def __init__(self, cutoff: int) -> None:
+        super().__init__(list(range(cutoff)), [Number([0], n) for n in range(cutoff)])
+        self._cutoff = cutoff
+
+    def probabilities(self, state, atol=1e-4):
+        fock_state = state.dm().to_fock(self._cutoff)
+        probs = math.astensor(
+            [
+                fock_state.representation.data[0][(ns * 2)]
+                for ns in product(self.meas_outcomes, repeat=len(state.modes))
+            ]
+        )
+        return self._validate_probs(probs, 1, atol)
+
+
+class HomodyneSampler(Sampler):
+    r"""
+    A sampler for homodyne measurements.
+
+    Args:
+        phi: The quadrature angle where ``0`` corresponds to ``x`` and ``\pi/2`` to ``p``.
+        bounds: The range of values to discretize over.
+        num: The number of points to discretize over.
+    """
+
+    def __init__(
+        self,
+        phi: float = 0,
+        bounds: tuple[float, float] = (-10, 10),
+        num: int = 1000,
```

Comment:
This too seems unnecessary. As for the PNR, it will work, but this should perhaps be set up as a continuous sampler. Maybe it's worth splitting the base sampler into a continuous and a discrete one. The discrete can be used for PNRs (and whatever discrete POVM a user wants to instantiate) and the continuous can be used for Homodyne Heterodyne etc..